src/mode/modeHandler.ts

```diff
@@ -129,50 +129,63 @@ export class VimSettings {
   useCtrlKeys     = false;
 }
 
+export enum SearchDirection {
+  Forward = 1,
+  Backward = -1
+}
+
 export class SearchState {
+  private static readonly MAX_SEARCH_RANGES = 1000;
+
   /**
    * Every range in the document that matches the search string.
    */
-  public matchRanges: vscode.Range[] = [];
+  private _matchRanges: vscode.Range[] = [];
+  public get matchRanges(): vscode.Range[] {
+    return this._matchRanges;
+  }
+
+  private _searchCursorStartPosition: Position;
+  public get searchCursorStartPosition(): Position {
+    return this._searchCursorStartPosition;
+  }
+
+  private _matchesDocVersion: number;
+  private _searchDirection: SearchDirection = SearchDirection.Forward;
 
   private _searchString = "";
   public get searchString(): string {
     return this._searchString;
   }
 
-  public searchCursorStartPosition: Position;
-
-  /**
-   * 1  === forward
-   * -1 === backward
-   */
-  public searchDirection = 1;
-
-
   public set searchString(search: string){
-    this._searchString = search;
-
-    this._recalculateSearchRanges();
+    if (this._searchString !== search) {
+      this._searchString = search;
+      this._recalculateSearchRanges(/*forceRecalc=*/true);
```

Comment:
I like the idea here, but I'd suggest to take it one step further and pass in `{ forceRecalc: true }`.